sopel/irc/modes.py

```diff
@@ -0,0 +1,234 @@
+"""Mode management for IRC channels.
+
+.. seealso::
+
+    https://modern.ircdocs.horse/#mode-message
+
+.. versionadded:: 8.0
+"""
+from __future__ import generator_stop
+
+from collections import namedtuple
+import logging
+from typing import Dict, Generator, List, Optional, Set, Tuple
+
+
+LOGGER = logging.getLogger(__name__)
+
+PARAM_ALWAYS = 'always'
+PARAM_ADDED = 'added'
+PARAM_REMOVED = 'removed'
```

Comment:
Is there such a thing? I only know of modes that always require a parameter; modes that need a parameter only when being set/added; and modes that forbid (never have) a parameter.

Sure you want to include code to handle this magical mode type that might never get real-world testing?

Because of this, I'm also wondering why the constructor for `ModeMessage` even needs to have the `type_param` argument passed in. Shouldn't it always be the same, because the types and their parameter requirements are standardized?

Reply:
The goal was to be able to handle any new type of modes, and for that I tried to designed `sopel.irc.modes` as a standalone library. At some point I'd love to just extract all that into its own package and distribute it as "here are some IRC tools for you". 😁 

So, for that to work, I made sure that one would be able to add mode types that are not the official ones (A, B, C, D): the spec allows it, even though servers should be very careful with those.